File: Code/CH11 Code/C#/Golf-c/CollectionClass.cs
using System;
using System.Collections;

namespace Golf_c
{
  /// <summary>
  /// Collections of courses and collections of scorecards
  /// </summary>
  /// 

  #region public enumerations

  public enum GolfClubs
  {
    One_iron,
    Two_Iron,
    Three_Iron,
    Four_Iron,
    Five_Iron,
    Six_Iron,
    Seven_Iron,
    Eight_Iron,
    Nine_Iron,
    PW,
    SW,
    GW,
    LW,
    Driver,
    Three_wood,
    Five_wood,
    Seven_wood,
    Nine_wood,
    Putter
  }

  public enum CoursePar
  {
    _27 = 27,
    _35 = 35,
    _36 = 36,
    _54 = 54,
    _70 = 70,
    _71 = 71,
    _72 = 72
  }

  public enum YardMarker
  {
    Red,
    White,
    Blue
  }

  #endregion

  public class Globals
  {
    public static int MinSlope{get{return 55;}}
    public static int MaxSlope{get{return 155;}}
    public static int SlopeLen
    {
      get
      {
        return(MaxSlope.ToString().Length > MinSlope.ToString().Length ? 
               MaxSlope.ToString().Length : MinSlope.ToString().Length);
      }
    }
    public static int NineHoles{get{return 9;}}
    public static int EighteenHoles{get{return 18;}}
  }


  //-----------------------------------------------------------------------------
  #region Courseinfo collection classes

  //This is a home grown strongly typed collection class and its object
  //Not only is it strongly typed but it allows me to hide some of the 
  //normal add/delete methods from other classes if need be.
  public class ICourseInfos : IEnumerable
  {
    //Slower than Hash table but more flexible
    //Can get item three ways
    //Most like VB type collection
    private SortedList      mCol;

    public ICourseInfos()
    {
      mCol = new  SortedList();
    }

    #region collection methods

    // enables foreach processing
    private mEnum GetEnumerator() 
    {
      return new mEnum(this);
    }

    //Property count
    public int Count
    {
      get { return mCol.Count; }
    }

    // ----- add method ------
    public void Add(ICourseInfo GolfCourse)
    {
      mCol.Add(GolfCourse.Name, GolfCourse);
    }

    // ----- overloaded remove method ------
    public void Remove(int Index)
    {
      mCol.RemoveAt(Index);
    }
    public void Remove(string key)
    {
      mCol.Remove(key);
    }

    // ----- overloaded item method ------
    public ICourseInfo Item(int index)
    {
      return (ICourseInfo) mCol.GetByIndex(index);
    }
    public ICourseInfo Item(string key)
    {
      return (ICourseInfo) mCol[key];
    }

    #endregion

    #region enumeration methods

    // Implement the GetEnumerator() method:
    IEnumerator IEnumerable.GetEnumerator() 
    {
      return GetEnumerator();
    }

    // Declare the enumerator and implement the IEnumerator interface:
    private class mEnum: IEnumerator 
    {
      private int nIndex;
      private ICourseInfos collection;

      // constructor. make the collection
      public mEnum(ICourseInfos coll) 
      {
        collection = coll;
        nIndex = -1;
      }

      // start over
      public void Reset() 
      {
        nIndex = -1;
      }

      // bump up the index
      public bool MoveNext() 
      {
        nIndex++;
        return(nIndex <  collection.mCol.Count);
      }

      // get the current object
      public ICourseInfo Current 
      {
        get {return (ICourseInfo) collection.mCol.GetByIndex(nIndex); }
      }

      // The current property on the IEnumerator interface:
      object IEnumerator.Current 
      {
        get { return(Current); }
      }

      #endregion
    }
  }

  public class ICourseInfo
  {
    #region locals

    private string    mName;
    private int       mLength;
    private int       mHoles;
    private int       mSlope;
    private CoursePar mPar;

    public ArrayList Hole = new ArrayList(); //Contains Tees objects
    public IScoreCardInfos  ScoreCards;

    #endregion

    public ICourseInfo()
    {
      mSlope = Globals.MinSlope;
      mName = "Undefined Course";
      ScoreCards = new IScoreCardInfos();
    }

    public ICourseInfo(string name)
    {
      mName = name;
      mSlope = Globals.MinSlope;
      NumberOfHoles = 9;
      ScoreCards = new IScoreCardInfos();
    }

    #region  Properties

    public string Name
    {
      get{return mName;}
      set{mName = value;}
    }

    public int Length
    {
      get
      {
        foreach(Tees Tee in Hole)
        {
          mLength += Tee.BlueDistance;
        }
        return mLength;
      }
 //     set{mLength = value;}
    }

    public int NumberOfHoles
    {
      get{return mHoles;}
      set
      {
        mHoles = value;
        Hole.Clear();
        for(int k=0; k<mHoles; k++)
        {
          Tees tee = new Tees();
          tee.HoleNumber = k+1;
          tee.BlueDistance = 0;
          tee.RedDistance = 0;
          tee.WhiteDistance = 0;
          Hole.Add(tee);
        }
        Hole.TrimToSize();
      }
    }

    public int Slope
    {
      get{return mSlope;}
      set{mSlope = value;}
    }

    public CoursePar Par
    {
      get{return mPar;}
      set{mPar = value;}
    }

    #endregion

    #region Methods

    public override string ToString()
    {
      return Name;
    }
    #endregion

  }

  public class Tees
  {
    private int mRedDistance;
    private int mBlueDistance;
    private int mWhiteDistance;
    private int mHoleNumber;
    private int mPar;

    public int RedDistance
    {
      //Only accept positive values
      get{return mRedDistance;}
      set{mRedDistance = value;}
    }
    public int BlueDistance
    {
      get{return mBlueDistance;}
      set{mBlueDistance = value;}
    }
    public int WhiteDistance
    {
      get{return mWhiteDistance;}
      set{mWhiteDistance = value;}
    }
    public int HoleNumber
    {
      get{return mHoleNumber;}
      set{mHoleNumber = value;}
    }
    public int Par
    {
      get{return mPar;}
      set{mPar = value;}
    }
  }

  #endregion

  //-----------------------------------------------------------------------------
  #region Scorecard collection classes

  public class IScoreCardInfos : IEnumerable
  {
    //Slower than Hash table but more flexible
    //Can get item three ways
    //Most like VB type collection
    private SortedList mCol;

    public IScoreCardInfos()
    {
      mCol = new  SortedList();
    }

    #region collection methods

    // enables foreach processing
    private mEnum GetEnumerator() 
    {
      return new mEnum(this);
    }

    //Property count
    public int Count
    {
      get { return mCol.Count; }
    }

    // ----- add method ------
    public void Add(IScoreCardInfo ScoreCard)
    {
      mCol.Add(ScoreCard.PlayDate, ScoreCard);
    }

    // ----- overloaded remove method ------
    public void Remove(int Index)
    {
      mCol.RemoveAt(Index);
    }
    public void Remove(string key)
    {
      mCol.Remove(key);
    }

    // ----- overloaded item method ------
    public IScoreCardInfo Item(int index)
    {
      return (IScoreCardInfo) mCol.GetByIndex(index);
    }
    public IScoreCardInfo Item(string key)
    {
      return (IScoreCardInfo) mCol[key];
    }

    #endregion

    #region enumeration methods

    // Implement the GetEnumerator() method:
    IEnumerator IEnumerable.GetEnumerator() 
    {
      return GetEnumerator();
    }

    // Declare the enumerator and implement the IEnumerator interface:
    private class mEnum: IEnumerator 
    {
      private int nIndex;
      private IScoreCardInfos collection;

      // constructor. make the collection
      public mEnum(IScoreCardInfos coll) 
      {
        collection = coll;
        nIndex = -1;
      }

      // start over
      public void Reset() 
      {
        nIndex = -1;
      }

      // bump up the index
      public bool MoveNext() 
      {
        nIndex++;
        return(nIndex <  collection.mCol.Count);
      }

      // get the current object
      public IScoreCardInfo Current 
      {
        get {return (IScoreCardInfo) collection.mCol.GetByIndex(nIndex); }
      }

      // The current property on the IEnumerator interface:
      object IEnumerator.Current 
      {
        get { return(Current); }
      }

      #endregion
    }
  }


  public class IScoreCardInfo
  {
    public IHoleDetailInfos holes;
    private DateTime mDate;

    public IScoreCardInfo()
    {
      holes = new IHoleDetailInfos();
      mDate = DateTime.Now;
    }

    public IScoreCardInfo(int numHoles)
    {
      holes = new IHoleDetailInfos();
      IHoleDetailInfo h;
      for(int k=0; k<numHoles; k++)
      {
        h = new IHoleDetailInfo();
        h.TotalShots = 0;
        h.Hole  = k+1;
        h.Putts = 0;
        holes.Add(h);
      }

      mDate = DateTime.Now;
    }

    public DateTime PlayDate
    {
      get{return mDate;}
      set{mDate = value;}
    }

    public int RoundScore
    {
      get
      {
        int score = 0;
        foreach(IHoleDetailInfo h in holes)
          score += h.TotalShots;
        return score;
      }
    }

  }

  #endregion

  //-----------------------------------------------------------------------------
  #region Hole Detail collection Classes

  public class IHoleDetailInfos : IEnumerable
  {
    //Slower than Hash table but more flexible
    //Can get item three ways
    //Most like VB type collection
    private SortedList mCol;

    public IHoleDetailInfos()
    {
      mCol = new  SortedList();
    }

    #region collection methods

    // enables foreach processing
    private mEnum GetEnumerator() 
    {
      return new mEnum(this);
    }

    //Property count
    public int Count
    {
      get { return mCol.Count; }
    }

    // ----- add method ------
    public void Add(IHoleDetailInfo hole)
    {
      mCol.Add(hole.ToString(), hole);
      mCol.TrimToSize();
    }

    // ----- overloaded remove method ------
    public void Remove(int Index)
    {
      mCol.RemoveAt(Index);
    }
    public void Remove(string key)
    {
      mCol.Remove(key);
    }

    // ----- overloaded item method ------
    public IHoleDetailInfo Item(int index)
    {
      return (IHoleDetailInfo) mCol.GetByIndex(index);
    }
    public IHoleDetailInfo Item(string key)
    {
      return (IHoleDetailInfo) mCol[key];
    }

    #endregion

    #region enumeration methods

    // Implement the GetEnumerator() method:
    IEnumerator IEnumerable.GetEnumerator() 
    {
      return GetEnumerator();
    }

    // Declare the enumerator and implement the IEnumerator interface:
    private class mEnum: IEnumerator 
    {
      private int nIndex;
      private IHoleDetailInfos collection;

      // constructor. make the collection
      public mEnum(IHoleDetailInfos coll) 
      {
        collection = coll;
        nIndex = -1;
      }

      // start over
      public void Reset() 
      {
        nIndex = -1;
      }

      // bump up the index
      public bool MoveNext() 
      {
        nIndex++;
        return(nIndex <  collection.mCol.Count);
      }

      // The current property on the IEnumerator interface:
      object IEnumerator.Current 
      {
        get { return(collection.mCol.GetByIndex(nIndex)); }
      }
    }

    #endregion
  }


  public class IHoleDetailInfo
  {
    #region Locals

    private YardMarker  mTeeBox;
    private int         mHole;
    private int         mPar;
    private GolfClubs   mTeeClub;
    private GolfClubs   mSecondClub;
    private bool        mHitfairway;
    private bool        mGood2Shot;
    private int         mShots2Green;
    private int         mPutts;
    private int         mTotalShots;

    #endregion

    public IHoleDetailInfo()
    {
      mTotalShots = 0;
    }

    public override string ToString()
    {
      return mHole.ToString();
    }

    #region Properties

    public YardMarker TeeBox
    {
      get{return mTeeBox;}
      set{mTeeBox = value;}
    }

    public GolfClubs ScondClub
    {
      get{return mSecondClub;}
      set{mSecondClub = value;}
    }

    public GolfClubs TeeClub
    {
      get{return mTeeClub;}
      set{mTeeClub = value;}
    }

    public bool GoodSecondShot
    {
      get{return mGood2Shot;}
      set{mGood2Shot = value;}
    }

    public bool HitFairway
    {
      get{return mHitfairway;}
      set{mHitfairway = value;}
    }

    public int TotalShots
    {
      get{return mTotalShots;}
      set{mTotalShots = value;}
    }

    public int Putts
    {
      get{return mPutts;}
      set{mPutts = value;}
    }

    public int ShotsToGreen
    {
      get{return mShots2Green;}
      set{mShots2Green = value;}
    }

    public bool GreenInReg
    {
      get{return((mPar - mShots2Green >= 2) ? true : false);}
    }

    public int Par
    {
      get{return mPar;}
      set{mPar = value;}
    }

    public int Hole
    {
      get{return mHole;}
      set{mHole = value;}
    }

    #endregion
  }

  #endregion

}

File: Code/CH11 Code/VB/Golf-vb/CollectionClass.vb

Imports System.Collections

#Region "public enumerations"

Public Enum GolfClubs
  One_iron
  Two_Iron
  Three_Iron
  Four_Iron
  Five_Iron
  Six_Iron
  Seven_Iron
  Eight_Iron
  Nine_Iron
  PW
  SW
  GW
  LW
  Driver
  Three_wood
  Five_wood
  Seven_wood
  Nine_wood
  Putter
End Enum

Public Enum CoursePar
  _27 = 27
  _35 = 35
  _36 = 36
  _54 = 54
  _70 = 70
  _71 = 71
  _72 = 72
End Enum

Public Enum YardMarker
  Red
  White
  Blue
End Enum

#End Region

Public Class Globals
  Public Shared ReadOnly Property MinSlope() As Int32
    Get
      Return 55
    End Get
  End Property
  Public Shared ReadOnly Property MaxSlope() As Int32
    Get
      Return 155
    End Get
  End Property
  Public Shared ReadOnly Property SlopeLen() As Int32
    Get
      Return IIf(MaxSlope.ToString.Length > MinSlope.ToString.Length, _
                 MaxSlope.ToString.Length, MinSlope.ToString.Length)
    End Get
  End Property
  Public Shared ReadOnly Property NineHoles() As Int32
    Get
      Return 9
    End Get
  End Property
  Public Shared ReadOnly Property EighteenHoles() As Int32
    Get
      Return 18
    End Get
  End Property
End Class

'-----------------------------------------------------------------------------
#Region "Courseinfo collection classes"

Public Class ICourseInfos
  Implements IEnumerable

  'Slower than Hash table but more flexible
  'Can get item three ways
  'Most like VB type collection
  Private mCol As SortedList

  Public Sub New()
    mCol = New SortedList()
  End Sub

  ' enables foreach processing
  Private Function GetMyEnumerator() As mEnum
    Return New mEnum(Me)
  End Function

  ' Implement the GetEnumerator() method:
  Public Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
    Return GetMyEnumerator()
  End Function

  'Property count
  Public ReadOnly Property Count() As Int32
    Get
      Return mCol.Count
    End Get
  End Property

  ' ----- add method ------
  Public Sub Add(ByVal card As ICourseInfo)
    mCol.Add(card.ToString(), card)
    mCol.TrimToSize()
  End Sub

  ' ----- overloaded remove method ------
  Public Sub Remove(ByVal Index As Int32)
    mCol.RemoveAt(Index)
  End Sub

  Public Sub Remove(ByVal key As String)
    mCol.Remove(key)
  End Sub

  ' ----- overloaded item method ------
  Public Function Item(ByVal index As Int32) As ICourseInfo
    Return CType(mCol.GetByIndex(index), ICourseInfo)
  End Function

  Public Function Item(ByVal key As String) As ICourseInfo
    Return CType(mCol(key), ICourseInfo)
  End Function

  ' Declare the enumerator and implement the IEnumerator interface:
  Private Class mEnum
    Implements IEnumerator

    Private nIndex As Int32
    Private collection As ICourseInfos

    ' constructor. make the collection
    Public Sub New(ByVal coll As ICourseInfos)
      collection = coll
      nIndex = -1
    End Sub

    ' start over
    Public Sub Reset() Implements IEnumerator.Reset
      nIndex = -1
    End Sub

    ' bump up the index
    Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
      nIndex += 1
      Return (nIndex < collection.mCol.Count)
    End Function

    ' get the current object
    Public ReadOnly Property GetCurrent() As ICourseInfo
      Get
        Return CType(collection.mCol.GetByIndex(nIndex), ICourseInfo)
      End Get
    End Property

    ' The current property on the IEnumerator interface:
    Public ReadOnly Property Current() As Object Implements IEnumerator.Current
      Get
        Return GetCurrent
      End Get
    End Property
  End Class

End Class


Public Class ICourseInfo

#Region "locals"

  Private mName As String
  Private mLength As Int32
  Private mHoles As Int32
  Private mSlope As Int32
  Private mPar As CoursePar

  Public Hole As ArrayList = New ArrayList()  'Contains Tees objects
  Public ScoreCards As IScoreCardInfos

#End Region

  Public Sub New()
    mSlope = Globals.MinSlope
    mName = "Undefined Course"
    ScoreCards = New IScoreCardInfos()
  End Sub

  Public Sub New(ByVal name As String)
    mName = name
    mSlope = Globals.MinSlope
    NumberOfHoles = 9
    ScoreCards = New IScoreCardInfos()
  End Sub

#Region "Properties"

  Public Property Name() As String
    Get
      Return mName
    End Get
    Set(ByVal Value As String)
      mName = Value
    End Set
  End Property

  Public ReadOnly Property Length() As Int32
    Get
      Dim Tee As Tees
      For Each Tee In Hole
        mLength += Tee.BlueDistance
      Next
      Return mLength
    End Get
    '     set{mLength = value;}
  End Property

  Public Property NumberOfHoles() As Int32
    Get
      Return mHoles
    End Get
    Set(ByVal Value As Int32)

      mHoles = Value
      Hole.Clear()
      Dim k As Int32
      Dim tee As Tees
      For k = 0 To mHoles - 1
        tee = New Tees()
        tee.HoleNumber = k + 1
        tee.BlueDistance = 0
        tee.RedDistance = 0
        tee.WhiteDistance = 0
        Hole.Add(tee)
      Next
      Hole.TrimToSize()
    End Set
  End Property

  Public Property Slope() As Int32
    Get
      Return mSlope
    End Get
    Set(ByVal Value As Int32)
      mSlope = Value
    End Set
  End Property

  Public Property Par() As CoursePar
    Get
      Return mPar
    End Get
    Set(ByVal Value As CoursePar)
      mPar = Value
    End Set
  End Property

#End Region

#Region "Methods"

  Public Overrides Function ToString() As String
    Return Name
  End Function

#End Region

End Class


Public Class Tees
  Private mRedDistance As Int32
  Private mBlueDistance As Int32
  Private mWhiteDistance As Int32
  Private mHoleNumber As Int32
  Private mPar As Int32

  Public Property RedDistance() As Int32
    'Only accept positive values
    Get
      Return mRedDistance
    End Get
    Set(ByVal Value As Int32)
      mRedDistance = Value
    End Set
  End Property

  Public Property BlueDistance() As Int32
    'Only accept positive values
    Get
      Return mBlueDistance
    End Get
    Set(ByVal Value As Int32)
      mBlueDistance = Value
    End Set
  End Property

  Public Property WhiteDistance() As Int32
    'Only accept positive values
    Get
      Return mWhiteDistance
    End Get
    Set(ByVal Value As Int32)
      mWhiteDistance = Value
    End Set
  End Property

  Public Property HoleNumber() As Int32
    'Only accept positive values
    Get
      Return mHoleNumber
    End Get
    Set(ByVal Value As Int32)
      mHoleNumber = Value
    End Set
  End Property

  Public Property Par() As Int32
    'Only accept positive values
    Get
      Return mPar
    End Get
    Set(ByVal Value As Int32)
      mPar = Value
    End Set
  End Property

End Class

#End Region

'-----------------------------------------------------------------------------
#Region "ScoreCard classes"

Public Class IScoreCardInfos
  Implements IEnumerable

  'Slower than Hash table but more flexible
  'Can get item three ways
  'Most like VB type collection
  Private mCol As SortedList

  Public Sub New()
    mCol = New SortedList()
  End Sub

  ' enables foreach processing
  Private Function GetMyEnumerator() As mEnum
    Return New mEnum(Me)
  End Function

  ' Implement the GetEnumerator() method:
  Public Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
    Return GetMyEnumerator()
  End Function

  'Property count
  Public ReadOnly Property Count() As Int32
    Get
      Return mCol.Count
    End Get
  End Property

  ' ----- add method ------
  Public Sub Add(ByVal ScoreCard As IScoreCardInfo)
    mCol.Add(ScoreCard.PlayDate, ScoreCard)
  End Sub

  ' ----- overloaded remove method ------
  Public Sub Remove(ByVal Index As Int32)
    mCol.RemoveAt(Index)
  End Sub

  Public Sub Remove(ByVal key As String)
    mCol.Remove(key)
  End Sub

  ' ----- overloaded item method ------
  Public Function Item(ByVal index As Int32) As IScoreCardInfo
    Return CType(mCol.GetByIndex(index), IScoreCardInfo)
  End Function

  Public Function Item(ByVal key As String) As IScoreCardInfo
    Return CType(mCol(key), IScoreCardInfo)
  End Function

  ' Declare the enumerator and implement the IEnumerator interface:
  Private Class mEnum
    Implements IEnumerator

    Private nIndex As Int32
    Private collection As IScoreCardInfos

    ' constructor. make the collection
    Public Sub New(ByVal coll As IScoreCardInfos)
      collection = coll
      nIndex = -1
    End Sub

    ' start over
    Public Sub Reset() Implements IEnumerator.Reset
      nIndex = -1
    End Sub

    ' bump up the index
    Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
      nIndex += 1
      Return (nIndex < collection.mCol.Count)
    End Function

    ' get the current object
    Public ReadOnly Property GetCurrent() As IScoreCardInfo
      Get
        Return CType(collection.mCol.GetByIndex(nIndex), IScoreCardInfo)
      End Get
    End Property

    ' The current property on the IEnumerator interface:
    Public ReadOnly Property Current() As Object Implements IEnumerator.Current
      Get
        Return GetCurrent
      End Get
    End Property
  End Class

End Class


Public Class IScoreCardInfo

  Public holes As IHoleDetailInfos
  Private mDate As DateTime

  Public Sub New()
    holes = New IHoleDetailInfos()
    mDate = DateTime.Now
  End Sub

  Public Sub New(ByVal numHoles As Int32)

    holes = New IHoleDetailInfos()
    Dim h As IHoleDetailInfo
    Dim k As Int32
    For k = 0 To numHoles - 1
      h = New IHoleDetailInfo()
      h.TotalShots = 0
      h.Hole = k + 1
      h.Putts = 0
      holes.Add(h)
    Next

    mDate = DateTime.Now
  End Sub

  Public Property PlayDate() As DateTime
    Get
      Return mDate
    End Get
    Set(ByVal Value As DateTime)
      mDate = Value
    End Set
  End Property

  Public ReadOnly Property RoundScore() As Int32
    Get
      Dim score As Int32 = 0
      Dim h As IHoleDetailInfo
      For Each h In holes
        score += h.TotalShots
      Next
      Return score
    End Get
  End Property

End Class

#End Region

'-----------------------------------------------------------------------------
#Region "Hole Detail collection Classes"

Public Class IHoleDetailInfos
  Implements IEnumerable

  'Slower than Hash table but more flexible
  'Can get item three ways
  'Most like VB type collection
  Private mCol As SortedList

  Public Sub New()
    mCol = New SortedList()
  End Sub

  ' enables foreach processing
  Private Function GetMyEnumerator() As mEnum
    Return New mEnum(Me)
  End Function

  ' Implement the GetEnumerator() method:
  Public Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
    Return GetMyEnumerator()
  End Function

  'Property count
  Public ReadOnly Property Count() As Int32
    Get
      Return mCol.Count
    End Get
  End Property

  ' ----- add method ------
  Public Sub Add(ByVal hole As IHoleDetailInfo)
    mCol.Add(hole.ToString(), hole)
    mCol.TrimToSize()
  End Sub

  ' ----- overloaded remove method ------
  Public Sub Remove(ByVal Index As Int32)
    mCol.RemoveAt(Index)
  End Sub

  Public Sub Remove(ByVal key As String)
    mCol.Remove(key)
  End Sub

  ' ----- overloaded item method ------
  Public Function Item(ByVal index As Int32) As IHoleDetailInfo
    Return CType(mCol.GetByIndex(index), IHoleDetailInfo)
  End Function

  Public Function Item(ByVal key As String) As IHoleDetailInfo
    Return CType(mCol(key), IHoleDetailInfo)
  End Function

  ' Declare the enumerator and implement the IEnumerator interface:
  Private Class mEnum
    Implements IEnumerator

    Private nIndex As Int32
    Private collection As IHoleDetailInfos

    ' constructor. make the collection
    Public Sub New(ByVal coll As IHoleDetailInfos)
      collection = coll
      nIndex = -1
    End Sub

    ' start over
    Public Sub Reset() Implements IEnumerator.Reset
      nIndex = -1
    End Sub

    ' bump up the index
    Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
      nIndex += 1
      Return (nIndex < collection.mCol.Count)
    End Function

    ' The current property on the IEnumerator interface:
    Public ReadOnly Property Current() As Object Implements IEnumerator.Current
      Get
        Return collection.mCol.GetByIndex(nIndex)
      End Get
    End Property
  End Class

End Class


Public Class IHoleDetailInfo

#Region "Locals"

  Private mTeeBox As YardMarker
  Private mHole As Int32
  Private mPar As Int32
  Private mTeeClub As GolfClubs
  Private mSecondClub As GolfClubs
  Private mHitfairway As Boolean
  Private mGood2Shot As Boolean
  Private mShots2Green As Int32
  Private mPutts As Int32
  Private mTotalShots As Int32

#End Region

  Public Sub New()
    mTotalShots = 0
  End Sub

  Public Overrides Function ToString() As String
    Return mHole.ToString()
  End Function

#Region "Properties"

  Public ReadOnly Property GreenInReg() As Boolean
    Get
      Return (IIf(mPar - mShots2Green >= 2, True, False))
    End Get
  End Property

  Public Property TeeBox() As YardMarker
    Get
      Return mTeeBox
    End Get
    Set(ByVal Value As YardMarker)
      mTeeBox = Value
    End Set
  End Property

  Public Property ScondClub() As GolfClubs
    Get
      Return mSecondClub
    End Get
    Set(ByVal Value As GolfClubs)
      mSecondClub = Value
    End Set
  End Property

  Public Property TeeClub() As GolfClubs
    Get
      Return mTeeClub
    End Get
    Set(ByVal Value As GolfClubs)
      mTeeClub = Value
    End Set
  End Property

  Public Property GoodSecondShot() As Boolean
    Get
      Return mGood2Shot
    End Get
    Set(ByVal Value As Boolean)
      mGood2Shot = Value
    End Set
  End Property

  Public Property HitFairway() As Boolean
    Get
      Return mHitfairway
    End Get
    Set(ByVal Value As Boolean)
      mHitfairway = Value
    End Set
  End Property

  Public Property TotalShots() As Int32
    Get
      Return mTotalShots
    End Get
    Set(ByVal Value As Int32)
      mTotalShots = Value
    End Set
  End Property

  Public Property Putts() As Int32
    Get
      Return mPutts
    End Get
    Set(ByVal Value As Int32)
      mPutts = Value
    End Set
  End Property

  Public Property ShotsToGreen() As Int32
    Get
      Return mShots2Green
    End Get
    Set(ByVal Value As Int32)
      mShots2Green = Value
    End Set
  End Property

  Public Property Par() As Int32
    Get
      Return mPar
    End Get
    Set(ByVal Value As Int32)
      mPar = Value
    End Set
  End Property

  Public Property Hole() As Int32
    Get
      Return mHole
    End Get
    Set(ByVal Value As Int32)
      mHole = Value
    End Set
  End Property

#End Region
End Class

#End Region
